server/channels/app/platform/web_hub.go

```diff
@@ -656,21 +689,85 @@ func (i *hubConnectionIndex) Remove(wc *WebConn) {
 	// get the conn slice.
 	userConnections := i.byUserId[wc.UserId]
 	// get the last connection.
 	last := userConnections[len(userConnections)-1]
-	// set the slot that we are trying to remove to be the last connection.
+	// https://go.dev/wiki/SliceTricks#delete-without-preserving-order
 	userConnections[userConnIndex] = last
-	// remove the last connection pointer from slice.
 	userConnections[len(userConnections)-1] = nil
-	// remove the last connection from the slice.
 	i.byUserId[wc.UserId] = userConnections[:len(userConnections)-1]
 	// set the index of the connection that was moved to the new index.
 	i.byConnection[last] = userConnIndex
 
+	connectionID := wc.GetConnectionID()
+	// Remove webconns from i.byChannelID
+	// This has O(n) complexity. We are trading off speed while removing
+	// a connection, to improve broadcasting a message.
+	for chID, webConns := range i.byChannelID {
+		// https://go.dev/wiki/SliceTricks#filtering-without-allocating
+		filtered := webConns[:0]
+		for _, conn := range webConns {
+			if conn.GetConnectionID() != connectionID {
+				filtered = append(filtered, conn)
+			}
+		}
+		for i := len(filtered); i < len(webConns); i++ {
+			webConns[i] = nil
+		}
+		i.byChannelID[chID] = filtered
+	}
+
 	delete(i.byConnection, wc)
-	delete(i.byConnectionId, wc.GetConnectionID())
+	delete(i.byConnectionId, connectionID)
+}
+
+func (i *hubConnectionIndex) InvalidateCMCacheForUser(userID string) {
+	// Clear out all user entries which belong to channels.
+	for chID, webConns := range i.byChannelID {
+		// https://go.dev/wiki/SliceTricks#filtering-without-allocating
+		filtered := webConns[:0]
+		for _, conn := range webConns {
+			if conn.UserId != userID {
+				filtered = append(filtered, conn)
+			}
+		}
+		for i := len(filtered); i < len(webConns); i++ {
+			webConns[i] = nil
+		}
+		i.byChannelID[chID] = filtered
+	}
+
+	// re-populate the cache
+	cm, err := i.getChannelMembersForUser(userID)
+	if err != nil {
+		panic(fmt.Errorf("error while getting GetAllChannelMembersForUser: %v", err))
+	}
+	for chID := range cm {
+		i.byChannelID[chID] = append(i.byChannelID[chID], i.ForUser(userID)...)
+	}
+}
+
+const numRetries = 2
+
+// getChannelMembersForUser is a wrapper around the store method which retries
+// for numRetries times. After that we panic, to fail loudly because it is even
+// worse to continue without the channel membership data.
+// Note: the hub has internal recovery mechanism to auto-restart in case of a panic.
+// So only one hub out of numCPU hubs will be affected.
+func (i *hubConnectionIndex) getChannelMembersForUser(userID string) (map[string]string, error) {
+	var cm map[string]string
+	var err error
+	for ind := 0; ind < numRetries; ind++ {
+		cm, err = i.store.Channel().GetAllChannelMembersForUser(request.EmptyContext(i.logger), userID, false, false)
+		if err == nil {
+			return cm, nil
+		}
+		// It is imperative that we don't block the hub
+		time.Sleep(time.Millisecond)
```

Comment:
Are we concerned about this millisecond here? Also, should we log the first error in the query in case it does affect performance?

Reply:
Not particularly concerned. It's a hit we need to take. Will log the first error. Thanks.

Reply:
Not sure I fully understand the use of `Sleep` here. The caller of `getChannelMembersForUser` is blocked regardless, so what exactly are we unblocking by sleeping?

Reply:
By letting the DB some time to recover from whatever it is failing from.

Reply:
I see. I find the comment a bit unclear if the purpose of the sleep is to let the DB "breath". Plus I am not sure one millisecond would help that much since we are talking about a couple of retries so if it failed a ms before, I say there's a good chance it could fail again. And now I wonder, how does the retry store layer fit into this? In that case we pause for 100ms which sounds more reasonable.

Reply:
The retry store layer is purely for retriable queries. e.g: ones which got cancelled due to a deadlock. Those are safe to be auto-retried. This is more about if the DB failed to reply due to load or something. So what you are suggesting is to remove the retry completely? I'd be fine with that.

Reply:
Yes, it could be simpler if we don't retry at all and fail fast. Otherwise, we'd have to bump the sleep to a more meaningful value, I feel, and block for longer as a result.

Reply:
Done.